src/main/ruby/truffleruby/core/splitter.rb

```diff
@@ -36,157 +36,225 @@
 
 module Truffle
   class Splitter
-    def self.split_characters(string, pattern, limit, tail_empty)
-      if limit
-        string.chars.take(limit - 1) << (string.size > (limit - 1) ? string[(limit - 1)..-1] : '')
-      else
-        ret = string.chars.to_a
-        # Use #byteslice because it returns the right class and taints
-        # automatically. This is just appending a "", which is this
-        # strange protocol if a negative limit is passed in
-        ret << string.byteslice(0,0) if tail_empty
-        ret
-      end
-    end
-
-    def self.valid_encoding?(string)
-      raise ArgumentError, "invalid byte sequence in #{string.encoding.name}" unless string.valid_encoding?
-    end
+    DEFAULT_PATTERN = ' '
 
-    def self.split(string, pattern, limit)
-      # Odd edge case
-      return [] if string.empty?
+    class << self
+      def split(string, pattern, limit, &block)
+        # Odd edge case
+        return result(string, [], &block) if string.empty?
 
-      tail_empty = false
+        tail_empty = false
 
-      if Primitive.undefined?(limit)
-        limited = false
-      else
-        limit = Primitive.rb_to_int limit
-
-        if limit > 0
-          return [string.dup] if limit == 1
-          limited = true
+        if Primitive.undefined?(limit)
+          limited = false
         else
-          if limit < 0
-            tail_empty = true
+          limit = Primitive.rb_to_int limit
+
+          if limit > 0
+            return [string.dup] if limit == 1
+            limited = true
+          else
+            if limit < 0
+              tail_empty = true
+            end
+            limited = false
```

Comment:
Would it be possible to combine `lim`, `limited`, `limit` to just be `limit` where `limit == nil` represents when it is not provided. It's okay to re-assign the  `limit` variable.

Reply:
Hm, let's try (I gonna try tomorrow) ... I didn't touch this Rubinius logic on purpose, but maybe it's time to reorganize / simplify this

Reply:
There was good advice (also found missed edge case and added test for it): https://github.com/oracle/truffleruby/pull/2052/commits/04fe83538b7604362f76f186aca95cee8db16411